Tc mode shift

Theodore-Chatziioannou opened the pull request:
Leg mode shift methods. When changing the mode for a specified leg, it will also change the mode of all legs within the same tour. There may be better ways to do this, so let me know if you can think of any.

Tests are placed withing test_2_activity_plan.py

KasiaKoz commented:
> there is one outstanding over there around defaulting to car if the person has access to a car or a driving licence if you want to have a go.

https://github.com/arup-group/pam/blob/tc_mode_shift/tests/test_5_policies.py#L726
@Theodore-Chatziioannou

Theodore-Chatziioannou commented on pam/activity.py:
```diff
@@ -624,13 +624,39 @@ def simplify_pt_trips(self):
+                for other_act in tour:
+                    #if any of the trip ends belongs in the tour change the mode
+                    if act_from.is_exact(other_act) or act_to.is_exact(other_act):
+                        plan.mode = target_mode
```
Agreed, we should change travel duration as well. I would like to try taking average travel time between the corresponding zones/modes for a given population? Unless it hits performance too much.

I guess we also need to regulate how the changed duration affects activities (ie work start/end time should probably stay fixed, with home activities more flexible, make sure the altered activity durations are still sensible etc). Do you think it would be better to do this through the modify/policy modules, or code it directly into the plan one?

fredshone replied:
Like what you're both thinking. `plan.mode_shift(seg, target_mode)` is in line with the previous pattern. By implementing methods at the plan level we can to all the resulting plan shuffling that might be required when shortening a duration. There is already some logic for doing this when a leg or activity is removed. I called it pivoting - basically expands all durations around the final home activity to fill the 24 hours. In most cases this shuffles everything forward.

I am nervous about a multiplier because i don't want to hard code any modes. The bloody spellings keep changing even within the TfL data sets. Theo's scrapiong of other plans is a fun way to go. I think the only viable way of doing this is to pre scrape O-D matrices and pass these to the mutator. We're ging to need to make this sort of stuff anyway. The times will be from the pre-change population - but still pretty good i think.

I guess an intermediate step is to pass a matrix of converters as kasia suggests, ie bus -> car = x0.8

fredshone replied:
for filtering by an attribute (ie if someone has a driving licence) - i would like this to work the same way as the existing policies. So that we would pass an attribute_filter to the policy

Theodore-Chatziioannou commented:
> Good stuff Theo, I really think we should also change duration when we change mode if possible though, or at least add some placeholders and todos in the code.
> I fixed the few tests that were failing, but more importantly, I deleted `xfail` tags from a couple of tests that move activity locations [commit](https://github.com/arup-group/pam/commit/97491239f08bdac0c8b63aa7b524f35a03fc143b) there is one outstanding over there around defaulting to car if the person has access to a car or a driving licence if you want to have a go.

Theodore-Chatziioannou commented:
@KasiaKoz re the licence check, would it be a good idea (or possible) to access the person's attributes through the Plan class? Or should I move the method to the Person or Policy classes?

KasiaKoz commented:
> @KasiaKoz re the licence check, would it be a good idea (or possible) to access the person's attributes through the Plan class? Or should I move the method to the Person or Policy classes?

on second though i don't think that is part of the mode shift scope, the policy methods that determine location shifts (`move_activity`) should just change the default when calling on `mode_shift`, so don't worry about it (in this PR anyway, if you feel like having a stab, by all means go for it)

Theodore-Chatziioannou commented:
> > @KasiaKoz re the licence check, would it be a good idea (or possible) to access the person's attributes through the Plan class? Or should I move the method to the Person or Policy classes?
> 
> on second though i don't think that is part of the mode shift scope, the policy methods that determine location shifts (`move_activity`) should just change the default when calling on `mode_shift`, so don't worry about it (in this PR anyway, if you feel like having a stab, by all means go for it)

A policy where x% of car licence holders switch from PT to car when commuting sounds quite relevant right now and a good example to demonstrate! Happy to have a go at it at some point..

Theodore-Chatziioannou commented:
@KasiaKoz @fredshone I have amended the mode shift methods to take into account speed differentials between modes. Vehicle speeds are passed in the parameters as a dictionary to determine leg durations, with default values derived from the National Travel Survey. 
Plan timestamps are then modified as follows: a) initially, downstream activities and legs are shifted in line with the revised travel times, b) home activities duration is adjusted to account for the change in the time spent travelling. It's reasonable to assume that non-home activities will not be affected in terms of duration. We may wish to change activity locations (ie closer when switching to walk), but I am assuming this will be applied at policy level.

Some tests related with the move_activity method were breaking after the last updates; I think the reason is mode shift within the method is meant to be used slightly differently. Please let me know if you have a view on this.

If you are happy with the updates, I can merge and build a mode shift policy scenario separately.
